__tests__/server.test.js

```diff
@@ -26,6 +26,47 @@ describe("server --> nc_news_test", () =>{
         });
     });
     
+    describe("GET /api", () =>{ 
+      test("Returns status code 200 and object with correct objects to a correct requests", () => {
+        return request(server)
+          .get("/api")
+          .expect(200)
+          .expect("Content-Type", /json/)
+          .then(({ body }) => {
+            const {endpoints} = body;
+            expect(endpoints).toBeInstanceOf(Object);
+            const allEndpoints = Object.keys(endpoints);
+            expect(allEndpoints.length).toBeGreaterThanOrEqual(1);
+            allEndpoints.forEach((endpoint)=>{
+              expect(endpoints[endpoint]).toBeInstanceOf(Object);
+              expect(Object.keys(endpoints[endpoint])).toHaveLength(4);
+              expect(typeof endpoints[endpoint].description).toBe("string");
+              expect(endpoints[endpoint].queries).toBeInstanceOf(Array);
+              expect(endpoints[endpoint].requestBody).toBeInstanceOf(Object);
+              expect(typeof endpoints[endpoint].description).toBe("string");
+            })
+          });
+      });
+      test("The response object correctly documents endpoints and responses", async() => {
+        const response = await request(server).get("/api");
+        const { endpoints } = response.body;
+        const allEndpoints = Object.keys(endpoints);
+        for (let i= 0; i<allEndpoints.length; i++){
+          const commandAndAddress = allEndpoints[i].split(" ");
+          if (commandAndAddress[0] === "GET") {
+            const endpointResponse = await request(server).get(
+              commandAndAddress[1]
+            );
+            const { body } = endpointResponse;
+            const { statusCode } = endpointResponse;
+            const expectedBody = endpoints[allEndpoints[i]].exampleResponse;
+            expect(statusCode).toBe(200);
+            expect(Object.keys(body)).toEqual(Object.keys(expectedBody));
+          }
+        }
+      }); 
+    });
+
```

Comment:
These tests for /api is fine, but also are getting quite complicated to read.

As we know that the endpoints.json file is what gets read in your model and served to the client as a response, but you may want to consider condensing this test into: 

- Require in the endpoints json file
- Assert that the response from `.get('/api')` is status code 200
- Assert that the response looks like the json required in.

Reply:
I'm not sure sure how to implement this, since response is served to the client from the DB, not from the endpoints.json file. It's unclear what does "Assert that the response looks like the json" mean in this case. Is comparing the keys enough? Or does the test seed data need to be synced with the endpoints.json file, and key values checked in the test? Doing so means introducing additional potential point of failure - in case if either test or endpoints.json data was changed, the person introducing the changes would have to remember to change the other file(s) as well, and be aware that they need changing in the first place. Which surpasses responses in endpoints.json being examples.

Or maybe the test needs to be written in another way entirely, that I am missing how to, the requirements were rather vague.